Random crash during compilation

I'm trying to integrate CLAD in my project, but I get a seemingly random crash when I do so.  CLAD is built from source

https://github.com/efeklisov/LiteRF-RadianceFields/blob/7934405fd2cbb58c12dc073a5aaba6439923bb01/example_tracer/example_tracer.cpp#L253

```
[ 25%] Building CXX object CMakeFiles/testapp.dir/example_tracer/example_tracer.cpp.o
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /usr/bin/clang++-17 -DUSE_ENZYME -D_CRT_SECURE_NO_WARNINGS -I/home/egorf/LiteRF -I/home/egorf/LiteRF/external/LiteMath -I/home/egorf/LiteRF/./external/nlohmann -I/home/egorf/LiteRF/./external -I/home/egorf/LiteRF/. -Wno-attributes -O0 -fopenmp -I /home/egorf/clad/inst/include -fplugin=/home/egorf/clad/inst/lib/clad.so -fplugin=/home/egorf/LiteRF/enzyme/build/Enzyme/ClangEnzyme-17.so -MD -MT CMakeFiles/testapp.dir/example_tracer/example_tracer.cpp.o -MF CMakeFiles/testapp.dir/example_tracer/example_tracer.cpp.o.d -o CMakeFiles/testapp.dir/example_tracer/example_tracer.cpp.o -c /home/egorf/LiteRF/example_tracer/example_tracer.cpp
1.      /home/egorf/LiteRF/example_tracer/example_tracer.cpp:260:1: current parser token 'void'
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  libLLVM-17.so.1    0x00007fbdadacc406 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 54
1  libLLVM-17.so.1    0x00007fbdadaca5b0 llvm::sys::RunSignalHandlers() + 80
2  libLLVM-17.so.1    0x00007fbdada20d70
3  libc.so.6          0x00007fbdac642520
4  libclang-cpp.so.17 0x00007fbdb5d9bf0b
5  libclang-cpp.so.17 0x00007fbdb5dbba3e clang::Sema::buildOverloadedCallSet(clang::Scope*, clang::Expr*, clang::UnresolvedLookupExpr*, llvm::MutableArrayRef<clang::Expr*>, clang::SourceLocation, clang::OverloadCandidateSet*, clang::ActionResult<clang::Expr*, true>*) + 110
6  clad.so            0x00007fbda9ef871a clad::DerivativeBuilder::noOverloadExists(clang::Expr*, llvm::MutableArrayRef<clang::Expr*>) + 488
7  clad.so            0x00007fbda9ef8e55 clad::DerivativeBuilder::BuildCallToCustomDerivativeOrNumericalDiff(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>> const&, llvm::SmallVectorImpl<clang::Expr*>&, clang::Scope*, clang::DeclContext*, bool, bool) + 1653
8  clad.so            0x00007fbdaa095d26 clad::ReverseModeVisitor::VisitCallExpr(clang::CallExpr const*) + 10510
9  clad.so            0x00007fbdaa07b7b8
10 clad.so            0x00007fbdaa078060
11 clad.so            0x00007fbdaa09cbd5 clad::ReverseModeVisitor::VisitImplicitCastExpr(clang::ImplicitCastExpr const*) + 89
12 clad.so            0x00007fbdaa07b940
13 clad.so            0x00007fbdaa078060
14 clad.so            0x00007fbdaa09b77e clad::ReverseModeVisitor::DifferentiateVarDecl(clang::VarDecl const*) + 2358
15 clad.so            0x00007fbdaa09c416 clad::ReverseModeVisitor::VisitDeclStmt(clang::DeclStmt const*) + 356
16 clad.so            0x00007fbdaa07a958
17 clad.so            0x00007fbdaa078060
18 clad.so            0x00007fbdaa09bfa7 clad::ReverseModeVisitor::DifferentiateSingleStmt(clang::Stmt const*, clang::Expr*) + 155
19 clad.so            0x00007fbdaa08f5fe clad::ReverseModeVisitor::VisitCompoundStmt(clang::CompoundStmt const*) + 338
20 clad.so            0x00007fbdaa07a8e8
21 clad.so            0x00007fbdaa078060
22 clad.so            0x00007fbdaa08deae clad::ReverseModeVisitor::DifferentiateWithClad() + 928
23 clad.so            0x00007fbdaa08cc6a clad::ReverseModeVisitor::Derive(clang::FunctionDecl const*, clad::DiffRequest const&) + 4158
24 clad.so            0x00007fbda9ef9776 clad::DerivativeBuilder::Derive(clad::DiffRequest const&) + 1606
25 clad.so            0x00007fbda9ee148f clad::plugin::CladPlugin::ProcessDiffRequest(clad::DiffRequest&) + 1383
26 clad.so            0x00007fbda9ee0c99 clad::plugin::CladPlugin::HandleTopLevelDecl(clang::DeclGroupRef) + 661
27 libclang-cpp.so.17 0x00007fbdb6d91d3c clang::MultiplexConsumer::HandleTopLevelDecl(clang::DeclGroupRef) + 44
28 libclang-cpp.so.17 0x00007fbdb50378ea clang::ParseAST(clang::Sema&, bool, bool) + 826
29 libclang-cpp.so.17 0x00007fbdb6d570c5 clang::FrontendAction::Execute() + 85
30 libclang-cpp.so.17 0x00007fbdb6cd26c4 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) + 708
31 libclang-cpp.so.17 0x00007fbdb6dd0e84 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) + 660
32 clang++-17         0x00005636d433c856 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) + 2454
33 clang++-17         0x00005636d433a3c5
34 libclang-cpp.so.17 0x00007fbdb6994c59
35 libLLVM-17.so.1    0x00007fbdada20b0c llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) + 140
36 libclang-cpp.so.17 0x00007fbdb699449e clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const + 366
37 libclang-cpp.so.17 0x00007fbdb695d40d clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const + 877
38 libclang-cpp.so.17 0x00007fbdb695d66e clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const + 142
39 libclang-cpp.so.17 0x00007fbdb69791ed clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) + 349
40 clang++-17         0x00005636d4339d2d clang_main(int, char**, llvm::ToolContext const&) + 11229
41 clang++-17         0x00005636d4346562 main + 50
42 libc.so.6          0x00007fbdac629d90
43 libc.so.6          0x00007fbdac629e40 __libc_start_main + 128
44 clang++-17         0x00005636d4336df5 _start + 37
clang++-17: error: clang frontend command failed with exit code 139 (use -v to see invocation)
Ubuntu clang version 17.0.6 (++20231209124227+6009708b4367-1~exp1~20231209124336.77)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
clang++-17: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang++-17: note: diagnostic msg: /tmp/example_tracer-cd5b81.cpp
clang++-17: note: diagnostic msg: /tmp/example_tracer-cd5b81.sh
clang++-17: note: diagnostic msg: 

********************
make[2]: *** [CMakeFiles/testapp.dir/build.make:90: CMakeFiles/testapp.dir/example_tracer/example_tracer.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/testapp.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```

--- a/include/clad/Differentiator/ReverseModeVisitor.h
+++ b/include/clad/Differentiator/ReverseModeVisitor.h
@@ -14,6 +14,7 @@
 #include "clad/Differentiator/ReverseModeVisitorDirectionKinds.h"
 #include "clad/Differentiator/VisitorBase.h"
 
+#include "clang/AST/Decl.h"
 #include "clang/AST/DeclCXX.h"
 #include "clang/AST/Expr.h"
 #include "clang/AST/ExprCXX.h"
@@ -29,6 +30,7 @@
 #include <array>
 #include <limits>
 #include <memory>
+#include <queue>
 #include <stack>
 #include <unordered_map>
 
@@ -397,6 +399,9 @@ namespace clad {
     StmtDiff VisitIntegerLiteral(const clang::IntegerLiteral* IL);
     StmtDiff VisitMemberExpr(const clang::MemberExpr* ME);
     StmtDiff VisitParenExpr(const clang::ParenExpr* PE);
+    StmtDiff VisitArrayInitLoopExpr(const clang::ArrayInitLoopExpr* AILE);
+    StmtDiff VisitArrayInitIndexExpr(const clang::ArrayInitIndexExpr* AIIE);
+    StmtDiff VisitOpaqueValueExpr(const clang::OpaqueValueExpr* OVE);
     virtual StmtDiff VisitReturnStmt(const clang::ReturnStmt* RS);
     StmtDiff VisitStmt(const clang::Stmt* S);
     virtual StmtDiff VisitUnaryOperator(const clang::UnaryOperator* UnOp);
@@ -718,6 +723,13 @@ namespace clad {
     void PopSwitchStmtInfo() { m_SwitchStmtsData.pop_back(); }
 
   private:
+    // When differentiating ArrayInitLoopExpr, we need to replace
+    // ArrayInitIndexExpr with real indices. We need to both add and pop them in
+    // the right order, so we use std::queue. For example, for `arr[i][j]`, we
+    // add `i` first, then `j`, and then pop them in the same order to generate
+    // the subscript expr.
+    std::queue<clang::VarDecl*> m_ArrayInitLoopIdx;
+
     // FIXME: This variable is used to track
     // whether we're currently visiting an init of a var decl.
     // This is only necessary because we don't create constructors
@@ -726,6 +738,7 @@ namespace clad {
     // other cases, we have to use InitListExpr and change the constructor
     // style. Remove this once we generate constructors explicitly.
     bool m_TrackVarDeclConstructor = false;
+
     /// A flag indicating if the Stmt is contained in a checkpointed loop.
     bool m_IsInsideCheckpointedLoop = false;
   };

--- a/include/clad/Differentiator/VisitorBase.h
+++ b/include/clad/Differentiator/VisitorBase.h
@@ -13,6 +13,7 @@
 
 #include "clang/AST/Expr.h"
 #include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/AST/Stmt.h"
 #include "clang/AST/StmtVisitor.h"
 #include "clang/AST/Type.h"
 #include "clang/Basic/Diagnostic.h"
@@ -28,6 +29,7 @@
 
 #include <array>
 #include <cassert>
+#include <cstddef>
 #include <stack>
 #include <unordered_map>
 
@@ -294,6 +296,14 @@ namespace clad {
     clang::Expr* BuildOperatorCall(clang::OverloadedOperatorKind OOK,
                                    llvm::MutableArrayRef<clang::Expr*> ArgExprs,
                                    clang::SourceLocation OpLoc = noLoc);
+
+    /// A shorthand to generage a standard loop of form
+    /// ```
+    /// for (type loopCounter = 0; loopCounter < N; ++loopCounter)
+    ///   body;
+    /// ```
+    clang::ForStmt* BuildStandardForLoop(clang::VarDecl* loopCounter, size_t N,
+                                         clang::Stmt* body);
     /// Function to resolve Unary Minus. If the leftmost operand
     /// has a Unary Minus then adds parens before adding the unary minus.
     /// \param[in] E Expression fed to the recursive call.

--- a/lib/Differentiator/DiffPlanner.cpp
+++ b/lib/Differentiator/DiffPlanner.cpp
@@ -709,6 +709,10 @@ static QualType GetDerivedFunctionType(const CallExpr* CE) {
           return false;
         return true;
       }
+      // The sub-stmt of OpaqueValueExpr is not visited automatically
+      bool VisitOpaqueValueExpr(const clang::OpaqueValueExpr* OVE) {
+        return TraverseStmt(OVE->getSourceExpr());
+      }
       // FIXME: This is a temporary measure until we add support for
       // `this` in varied analysis.
       bool VisitCXXThisExpr(const clang::CXXThisExpr* TE) { return false; }

--- a/lib/Differentiator/ReverseModeVisitor.cpp
+++ b/lib/Differentiator/ReverseModeVisitor.cpp
@@ -723,6 +723,49 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
     return Visit(ILE->getSubExpr(), dfdx());
   }
 
+  StmtDiff
+  ReverseModeVisitor::VisitArrayInitLoopExpr(const ArrayInitLoopExpr* AILE) {
+    // Since ArrayInitLoopExpr is not possible to express with regular syntax,
+    // we have to replicate it with loops.
+    // The code we're differentiated is of the form
+    // res = ArrayInitLoopExpr(arr[ArrayInitIndexExpr])
+    // We have to replace ArrayInitIndexExpr with an actual index `i`
+    // and wrap the code in a for loop to compute the derivative as follows:
+    // for (int i = 0; i < N; ++i)
+    //   _d_arr[i] += _d_res[i];
+    beginScope(Scope::DeclScope);
+    VarDecl* idxDecl = BuildVarDecl(m_Context.UnsignedIntTy, "i",
+                                    getZeroInit(m_Context.IntTy));
+    // Push the index to the queue so that we can replace ArrayInitIndexExpr
+    // when we encounter it.
+    m_ArrayInitLoopIdx.push(idxDecl);
+    Expr* idx = BuildDeclRef(idxDecl);
+    // Build `_d_res[i]`
+    Expr* diff = BuildArraySubscript(dfdx(), {idx});
+    beginBlock(direction::reverse);
+    Visit(AILE->getSubExpr(), diff);
+    Stmt* block = utils::unwrapIfSingleStmt(endBlock(direction::reverse));
+    Stmt* loopDiff = BuildStandardForLoop(
+        idxDecl, AILE->getArraySize().getZExtValue(), block);
+    addToCurrentBlock(loopDiff, direction::reverse);
+    endScope();
+    // We cannot clone ArrayInitLoopExpr because it's not possible to express
+    // with standard c++ syntax.
+    return {};
+  }
+
+  StmtDiff
+  ReverseModeVisitor::VisitArrayInitIndexExpr(const ArrayInitIndexExpr* AIIE) {
+    VarDecl* idxDecl = m_ArrayInitLoopIdx.front();
+    m_ArrayInitLoopIdx.pop();
+    return {BuildDeclRef(idxDecl)};
+  }
+
+  StmtDiff
+  ReverseModeVisitor::VisitOpaqueValueExpr(const OpaqueValueExpr* OVE) {
+    return Visit(OVE->getSourceExpr(), dfdx());
+  }
+
   StmtDiff ReverseModeVisitor::VisitStmt(const Stmt* S) {
     diagUnsupported(S);
     // Unknown stmt, just clone it.

--- a/lib/Differentiator/VisitorBase.cpp
+++ b/lib/Differentiator/VisitorBase.cpp
@@ -23,6 +23,7 @@
 #include "clang/AST/Expr.h"
 #include "clang/AST/NestedNameSpecifier.h"
 #include "clang/AST/OperationKinds.h"
+#include "clang/AST/Stmt.h"
 #include "clang/AST/TemplateBase.h"
 #include "clang/Basic/OperatorKinds.h"
 #include "clang/Basic/SourceManager.h"
@@ -41,6 +42,7 @@
 #include "llvm/Support/Casting.h"
 
 #include <algorithm>
+#include <cstddef>
 #include <numeric>
 
 #include "clad/Differentiator/Compatibility.h"
@@ -262,6 +264,17 @@ namespace clad {
   DeclStmt* VisitorBase::BuildDeclStmt(llvm::MutableArrayRef<Decl*> Decls) {
     auto DGR = DeclGroupRef::Create(m_Context, Decls.data(), Decls.size());
     return new (m_Context) DeclStmt(DGR, noLoc, noLoc);
+  }
+
+  ForStmt* VisitorBase::BuildStandardForLoop(VarDecl* loopCounter, size_t N,
+                                             Stmt* body) {
+    Stmt* init = BuildDeclStmt(loopCounter);
+    Expr* numExpr =
+        ConstantFolder::synthesizeLiteral(m_Context.IntTy, m_Context, N);
+    Expr* cond = BuildOp(BO_LT, BuildDeclRef(loopCounter), numExpr);
+    Expr* inc = BuildOp(UO_PreInc, BuildDeclRef(loopCounter));
+    return new (m_Context) ForStmt(m_Context, init, cond, /*CondVar=*/nullptr,
+                                   inc, body, noLoc, noLoc, noLoc);
   }
 
   DeclRefExpr* VisitorBase::BuildDeclRef(DeclaratorDecl* D,

--- a/test/Gradient/Constructors.C
+++ b/test/Gradient/Constructors.C
@@ -445,6 +445,123 @@ double fn9(S& s) {
 // CHECK-NEXT:      }
 // CHECK-NEXT:  }
 
+struct arrWrapper {
+    double arr[2];
+};
+
+// CHECK:  static inline constexpr void constructor_pullback(const arrWrapper &arg, arrWrapper *_d_this, arrWrapper *_d_arg) noexcept {
+// CHECK-NEXT:      for (unsigned {{int|long}} i = 0; i < 2; ++i)
+// CHECK-NEXT:          (*_d_arg).arr[i] += _d_this->arr[i];
+// CHECK-NEXT:  }
+
+double fn10(double x, double y) {
+    arrWrapper a = {x, y};
+    arrWrapper b = a;
+    return b.arr[0] + b.arr[1];
+}
+
+// CHECK:  void fn10_grad(double x, double y, double *_d_x, double *_d_y) {
+// CHECK-NEXT:      arrWrapper _d_a = {{.*0.*}};
+// CHECK-NEXT:      arrWrapper a = {{.*x, y.*}};
+// CHECK-NEXT:      arrWrapper _d_b = _d_a;
+// CHECK-NEXT:      arrWrapper b = a;
+// CHECK-NEXT:      {
+// CHECK-NEXT:          _d_b.arr[0] += 1;
+// CHECK-NEXT:          _d_b.arr[1] += 1;
+// CHECK-NEXT:      }
+// CHECK-NEXT:      arrWrapper::constructor_pullback(a, &_d_b, &_d_a);
+// CHECK-NEXT:      {
+// CHECK-NEXT:          *_d_x += _d_a.arr[0];
+// CHECK-NEXT:          *_d_y += _d_a.arr[1];
+// CHECK-NEXT:      }
+// CHECK-NEXT:  }
+
+struct arr2DWrapper {
+    double arr[1][2];
+};
+
+// CHECK:  static inline constexpr void constructor_pullback(const arr2DWrapper &arg, arr2DWrapper *_d_this, arr2DWrapper *_d_arg) noexcept {
+// CHECK-NEXT:      for (unsigned {{int|long}} i = 0; i < 1; ++i)
+// CHECK-NEXT:          for (unsigned {{int|long}} i0 = 0; i0 < 2; ++i0)
+// CHECK-NEXT:              (*_d_arg).arr[i][i0] += _d_this->arr[i][i0];
+// CHECK-NEXT:  }
+
+double fn11(double x, double y) {
+    arr2DWrapper a = {x, y};
+    arr2DWrapper b = a;
+    return b.arr[0][0] + b.arr[0][1];
+}
+
+// CHECK:  void fn11_grad(double x, double y, double *_d_x, double *_d_y) {
+// CHECK-NEXT:      arr2DWrapper _d_a = {{.*0.*}};
+// CHECK-NEXT:      arr2DWrapper a = {{.*x, y.*}};
+// CHECK-NEXT:      arr2DWrapper _d_b = _d_a;
+// CHECK-NEXT:      arr2DWrapper b = a;
+// CHECK-NEXT:      {
+// CHECK-NEXT:          _d_b.arr[0][0] += 1;
+// CHECK-NEXT:          _d_b.arr[0][1] += 1;
+// CHECK-NEXT:      }
+// CHECK-NEXT:      arr2DWrapper::constructor_pullback(a, &_d_b, &_d_a);
+// CHECK-NEXT:      {
+// CHECK-NEXT:          *_d_x += _d_a.arr[0][0];
+// CHECK-NEXT:          *_d_y += _d_a.arr[0][1];
+// CHECK-NEXT:      }
+// CHECK-NEXT:  }
+
+struct cust_double {
+    cust_double(double x = 0): val(x) {} 
+    double val;
+};
+
+// CHECK:  static void constructor_pullback(double x, cust_double *_d_this, double *_d_x) {
+// CHECK-NEXT:      {
+// CHECK-NEXT:          *_d_x += _d_this->val;
+// CHECK-NEXT:          _d_this->val = 0.;
+// CHECK-NEXT:      }
+// CHECK-NEXT:  }
+
+// CHECK:  static inline constexpr void constructor_pullback(const cust_double &arg, cust_double *_d_this, cust_double *_d_arg) noexcept {
+// CHECK-NEXT:      {
+// CHECK-NEXT:          (*_d_arg).val += _d_this->val;
+// CHECK-NEXT:          _d_this->val = 0.;
+// CHECK-NEXT:      }
+// CHECK-NEXT:  }
+
+struct arrStructWrapper {
+    cust_double arr[2];
+};
+
+// CHECK:  static inline constexpr void constructor_pullback(const arrStructWrapper &arg, arrStructWrapper *_d_this, arrStructWrapper *_d_arg) noexcept {
+// CHECK-NEXT:      for (unsigned {{int|long}} i = 0; i < 2; ++i)
+// CHECK-NEXT:          cust_double::constructor_pullback(arg.arr[i], &_d_this->arr[i], &(*_d_arg).arr[i]);
+// CHECK-NEXT:  }
+
+double fn12(double x, double y) {
+    arrStructWrapper a = {x, y};
+    arrStructWrapper b = a;
+    return b.arr[0].val + b.arr[1].val;
+}
+
+// CHECK:  void fn12_grad(double x, double y, double *_d_x, double *_d_y) {
+// CHECK-NEXT:      arrStructWrapper _d_a = {{.*}};
+// CHECK-NEXT:      arrStructWrapper a = {{.*x, y.*}};
+// CHECK-NEXT:      arrStructWrapper _d_b = _d_a;
+// CHECK-NEXT:      arrStructWrapper b = a;
+// CHECK-NEXT:      {
+// CHECK-NEXT:          _d_b.arr[0].val += 1;
+// CHECK-NEXT:          _d_b.arr[1].val += 1;
+// CHECK-NEXT:      }
+// CHECK-NEXT:      arrStructWrapper::constructor_pullback(a, &_d_b, &_d_a);
+// CHECK-NEXT:      {
+// CHECK-NEXT:          double _r0 = 0.;
+// CHECK-NEXT:          cust_double::constructor_pullback(x, &_d_a.arr[0], &_r0);
+// CHECK-NEXT:          *_d_x += _r0;
+// CHECK-NEXT:          double _r1 = 0.;
+// CHECK-NEXT:          cust_double::constructor_pullback(y, &_d_a.arr[1], &_r1);
+// CHECK-NEXT:          *_d_y += _r1;
+// CHECK-NEXT:      }
+// CHECK-NEXT:  }
+
 int main() {
     double d_i, d_j;
 
@@ -474,6 +591,14 @@ int main() {
     S s{new double[3]{5, 6, 7}}, _d_s{new double[3]{0}};
     auto dfn9 = clad::gradient(fn9);
     dfn9.execute(s, &_d_s);
-    printf("{%.2f, %.2f, %.2f}\n", _d_s.a[0], _d_s.a[1], _d_s.a[2]);
-    // TEST_GRADIENT(fn9, /*numOfDerivativeArgs=*/1, s, &d_s);    // CHECK-EXEC: {0.00, 0.00, 1.00}
+    printf("{%.2f, %.2f, %.2f}\n", _d_s.a[0], _d_s.a[1], _d_s.a[2]);   // CHECK-EXEC: {0.00, 0.00, 1.00}
+
+    INIT_GRADIENT(fn10);
+    TEST_GRADIENT(fn10, /*numOfDerivativeArgs=*/2, 7, 2, &d_i, &d_j);    // CHECK-EXEC: {1.00, 1.00}
+
+    INIT_GRADIENT(fn11);
+    TEST_GRADIENT(fn11, /*numOfDerivativeArgs=*/2, 9, -1, &d_i, &d_j);    // CHECK-EXEC: {1.00, 1.00}
+
+    INIT_GRADIENT(fn12);
+    TEST_GRADIENT(fn12, /*numOfDerivativeArgs=*/2, 3, 6, &d_i, &d_j);    // CHECK-EXEC: {1.00, 1.00}
 }

